src/storage/storage_structure/lists/rel_update_store.cpp

```diff
@@ -0,0 +1,302 @@
+                    .at(boundNodeID.offset)
+                    ->deletedRelIDs.insert(relID);
+            }
+        }
+    }
+}
+
+uint64_t RelUpdateStore::getNumInsertedRelsForNodeOffset(
+    ListFileID& listFileID, node_offset_t nodeOffset) const {
+    auto relUpdatesForNodeOffset = getRelUpdatesForNodeOffsetIfExists(listFileID, nodeOffset);
+    if (relUpdatesForNodeOffset == nullptr) {
+        return 0;
+    }
+    return relUpdatesForNodeOffset->insertedRelsTupleIdxInFT.size();
+}
+
+void RelUpdateStore::readValues(
+    ListFileID& listFileID, ListHandle& listHandle, shared_ptr<ValueVector> valueVector) const {
+    auto numTuplesToRead = listHandle.getNumValuesToRead();
+    auto nodeOffset = listHandle.getBoundNodeOffset();
+    if (numTuplesToRead == 0) {
+        valueVector->state->initOriginalAndSelectedSize(0);
+        return;
+    }
+    auto vectorsToRead = vector<shared_ptr<ValueVector>>{valueVector};
+    auto columnsToRead = vector<ft_col_idx_t>{getColIdxInFT(listFileID)};
+    auto relNodeTableAndDir = getRelNodeTableAndDirFromListFileID(listFileID);
+    auto& listUpdates = relUpdatesPerTablePerDirection[relNodeTableAndDir.dir]
+                            .at(relNodeTableAndDir.srcNodeTableID)
+                            .at(StorageUtils::getListChunkIdx(nodeOffset))
+                            .at(nodeOffset);
+    insertedRels->lookup(vectorsToRead, columnsToRead, listUpdates->insertedRelsTupleIdxInFT,
+        listHandle.getStartElemOffset(), numTuplesToRead);
+    valueVector->state->originalSize = numTuplesToRead;
+}
+
+bool RelUpdateStore::hasAnyDeletedRelsInPersistentStore(
+    ListFileID& listFileID, node_offset_t nodeOffset) const {
+    auto relUpdatesForNodeOffset = getRelUpdatesForNodeOffsetIfExists(listFileID, nodeOffset);
+    if (relUpdatesForNodeOffset == nullptr) {
+        return false;
+    }
+    return !relUpdatesForNodeOffset->deletedRelIDs.empty();
+}
+
+void RelUpdateStore::updateRelIfNecessary(const shared_ptr<ValueVector>& srcNodeIDVector,
+    const shared_ptr<ValueVector>& dstNodeIDVector, const RelUpdateInfo& relUpdateInfo) {
+    auto srcNodeID = srcNodeIDVector->getValue<nodeID_t>(
+        srcNodeIDVector->state->selVector->selectedPositions[0]);
+    auto dstNodeID = dstNodeIDVector->getValue<nodeID_t>(
+        dstNodeIDVector->state->selVector->selectedPositions[0]);
+    bool insertUpdatedRel = true;
+    for (auto direction : REL_DIRECTIONS) {
+        auto boundNodeID = direction == FWD ? srcNodeID : dstNodeID;
+        if (relUpdatesPerTablePerDirection[direction].contains(boundNodeID.tableID)) {
+            if (insertUpdatedRel) {
+                updatedRels[relUpdateInfo.propertyID]->append(
+                    vector<shared_ptr<ValueVector>>{relUpdateInfo.propertyVector});
+                insertUpdatedRel = false;
+            }
+            auto& relUpdatesPerNodeOffset = relUpdatesPerTablePerDirection[direction].at(
```

Comment:
Added a new function:
```
    ListsUpdatesForNodeOffset* getOrCreateListsUpdatesForNodeOffset(
        RelDirection relDirection, nodeID_t nodeID);
```
It returns the listsUpdatesForNodeOffset if exists. If not exists, it will insert a new listsUpdatesForNodeOffset.